File: 30-Day LeetCoding Challenge/Python/findMaxLength.py
class Solution(object):
    def findMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        dic = {0: 0}
        ans = 0
        sum = 0
        for i in range(len(nums)):
            if nums[i] == 0:
                sum += -1
            else:
                sum += 1
            if sum in dic:
                ans = max(ans, i + 1 - dic[sum])
            else:
                dic[sum] = i + 1
        return ans

s = Solution()
input = [0,1,0,0,1,0,1,1,1,0,0,0]
print(s.findMaxLength(input))

File: 30-Day LeetCoding Challenge/C++/findMaxLength.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include<unordered_map> 
using namespace std;


class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int sum = 0, ans = 0;
        unordered_map<int, int> m;
        m[0] = 0;
        
        for(int i = 0; i < nums.size(); i++) {
            if(nums[i] == 1) sum += 1;
            else sum += -1;
            auto it = m.find(sum);
            if(it != m.end()) ans = max(ans, i+1-m[sum]);
            else m[sum] = i+1;
        }
        return ans;
    }
};
int main() {
    vector<int> input = {0,1,0,0,1,0,1,1,1,0,0,0};
    Solution s;
    cout << s.findMaxLength(input);
    return 0;
}